lib/pure/concurrency/threadpool.nim

```diff
@@ -234,43 +236,42 @@ proc awaitAndThen*[T](fv: FlowVar[T]; action: proc (x: T) {.closure.}) =
   ## Note that due to Nim's parameter passing semantics this
   ## means that ``T`` doesn't need to be copied so ``awaitAndThen`` can
   ## sometimes be more efficient than `^ proc <#^,FlowVar[T]>`_.
-  blockUntil(fv)
-  when T is string or T is seq:
+  blockUntil(fv[])
+  when defined(nimV2):
+    action(fv.blob)
+  elif T is string or T is seq:
     action(cast[T](fv.data))
   elif T is ref:
     {.error: "'awaitAndThen' not available for FlowVar[ref]".}
   else:
     action(fv.blob)
-  finished(fv)
+  finished(fv[])
 
 proc unsafeRead*[T](fv: FlowVar[ref T]): ptr T =
   ## Blocks until the value is available and then returns this value.
-  blockUntil(fv)
-  result = cast[ptr T](fv.data)
-  finished(fv)
-
-proc `^`*[T](fv: FlowVar[ref T]): ref T =
-  ## Blocks until the value is available and then returns this value.
-  blockUntil(fv)
-  let src = cast[ref T](fv.data)
+  blockUntil(fv[])
   when defined(nimV2):
-    result = src
+    result = cast[ptr T](fv.blob)
   else:
-    deepCopy result, src
-  finished(fv)
+    result = cast[ptr T](fv.data)
+  finished(fv[])
 
-proc `^`*[T](fv: FlowVar[T]): T =
-  ## Blocks until the value is available and then returns this value.
-  blockUntil(fv)
-  when T is string or T is seq:
-    let src = cast[T](fv.data)
-    when defined(nimV2):
-      result = src
-    else:
-      deepCopy result, src
-  else:
+when defined(nimV2):
+  proc `^`*[T](fv: FlowVar[T]): T =
+    ## Blocks until the value is available and then returns this value.
+    blockUntil(fv[])
     result = fv.blob
-  finished(fv)
+    finished(fv[])
+
+else:
+  proc `^`*[T](fv: FlowVar[T]): T =
+    ## Blocks until the value is available and then returns this value.
+    blockUntil(fv[])
+    when T is string or T is seq or T is ref:
```

Comment:
Why not get rid of the 2 different procs in each when/else branch and put `when defined(nimV2) and (T is string or T is seq or T is ref):` here?

Reply:
done